source/grid/grid_tools.cc

```diff
@@ -429,6 +429,140 @@ namespace GridTools
 
 
 
+  // Generic functions for appending face data in 2D or 3D. TODO: we can
+  // remove these once we have 'if constexpr'.
+  namespace
+  {
+    void
+    append_face_data(const CellData<0> & /*face_data*/,
+                     SubCellData & /*subcell_data*/)
+    {
+      Assert(false, ExcInternalError());
+    }
+
+    void
+    append_face_data(const CellData<1> &face_data, SubCellData &subcell_data)
+    {
+      subcell_data.boundary_lines.push_back(face_data);
+    }
+
+    void
+    append_face_data(const CellData<2> &face_data, SubCellData &subcell_data)
+    {
+      subcell_data.boundary_quads.push_back(face_data);
+    }
+  } // namespace
+
+
+
+  template <int dim, int spacedim>
+  std::
+    tuple<std::vector<Point<spacedim>>, std::vector<CellData<dim>>, SubCellData>
+    get_coarse_mesh_description(const Triangulation<dim, spacedim> &tria)
+  {
+    Assert(1 <= tria.n_levels(),
+           ExcMessage("The input triangulation must be non-empty."));
+
+    std::tuple<std::vector<Point<spacedim>>,
+               std::vector<CellData<dim>>,
+               SubCellData>
+      result;
+
+    std::vector<Point<spacedim>> &vertices     = std::get<0>(result);
+    std::vector<CellData<dim>> &  cells        = std::get<1>(result);
+    SubCellData &                 subcell_data = std::get<2>(result);
+
+    // Set up a one-to-one and ascending numbering from the current level 0
+    // vertices (which may not be contiguous) to a new, contiguous set of
```

Comment:
Is this actually true? I mean your code below isn't going to be wrong, but I think that the level 0 vertices must be contiguous. At least I can't conjure up any scenarios where that wouldn't be the case.

Reply:
The thing is that we *create* the triangulation on level 0, so these are exactly the vertices that are created before all others. And because we have no way of getting rid of coarse level cells, there is also no way to create holes in the numbering of coarse level vertices later on.

Reply:
I went back and got rid of the *de facto* identity mapping. I did add an assertion to verify that we actually use all vertices, just in case.